be/src/exec/pipeline/sort/partition_sort_sink_operator.cpp

```diff
@@ -41,6 +41,11 @@ Status PartitionSortSinkOperator::push_chunk(RuntimeState* state, const vectoriz
 }
 
 Status PartitionSortSinkOperator::set_finishing(RuntimeState* state) {
+    // skip sorting if cancelled
+    if (state->is_cancelled()) {
+        _is_finished = true;
+        return Status::Cancelled("runtime state is cancelled");
```

Comment:
Should we return `Status::OK()` here? Because, BE will consider there is internal error and report error status to BE again, if `set_finishing()` returns non-OK status.
We could continue to deal with cancel logic of the next operators, if we return `Status::OK()`.


```C++
// PipelineDriver::process
RETURN_IF_ERROR(_mark_operator_finishing(next_op, runtime_state));

// GlobalDriverExecutor::_worker_thread
auto status = driver->process(runtime_state, worker_id);
if (!status.ok() || is_big_query) {
    if (is_big_query) {
        LOG(WARNING) << "[Driver] Process exceed limit, query_id="
                     << print_id(driver->query_ctx()->query_id())
                     << ", instance_id=" << print_id(driver->fragment_ctx()->fragment_instance_id());
        query_ctx->cancel(Status::Corruption("exceed limit, is big query"));
    } else {
        LOG(WARNING) << "[Driver] Process error, query_id=" << print_id(driver->query_ctx()->query_id())
                     << ", instance_id=" << print_id(driver->fragment_ctx()->fragment_instance_id())
                     << ", error=" << status.status().to_string();
        query_ctx->cancel(status.status());
    }
    driver->cancel_operators(runtime_state);
    if (driver->is_still_pending_finish()) {
        driver->set_driver_state(DriverState::PENDING_FINISH);
        _blocked_driver_poller->add_blocked_driver(driver);
    } else {
        _finalize_driver(driver, runtime_state, DriverState::INTERNAL_ERROR);
    }
    continue;
}
```

Reply:
> Should we return `Status::OK()` here? Because, BE will consider there is internal error and report error status to BE again, if `set_finishing()` returns non-OK status. We could continue to deal with cancel logic of the next operators, if we return `Status::OK()`.
> 
> ```c++
> // PipelineDriver::process
> RETURN_IF_ERROR(_mark_operator_finishing(next_op, runtime_state));
> 
> // GlobalDriverExecutor::_worker_thread
> auto status = driver->process(runtime_state, worker_id);
> if (!status.ok() || is_big_query) {
>     if (is_big_query) {
>         LOG(WARNING) << "[Driver] Process exceed limit, query_id="
>                      << print_id(driver->query_ctx()->query_id())
>                      << ", instance_id=" << print_id(driver->fragment_ctx()->fragment_instance_id());
>         query_ctx->cancel(Status::Corruption("exceed limit, is big query"));
>     } else {
>         LOG(WARNING) << "[Driver] Process error, query_id=" << print_id(driver->query_ctx()->query_id())
>                      << ", instance_id=" << print_id(driver->fragment_ctx()->fragment_instance_id())
>                      << ", error=" << status.status().to_string();
>         query_ctx->cancel(status.status());
>     }
>     driver->cancel_operators(runtime_state);
>     if (driver->is_still_pending_finish()) {
>         driver->set_driver_state(DriverState::PENDING_FINISH);
>         _blocked_driver_poller->add_blocked_driver(driver);
>     } else {
>         _finalize_driver(driver, runtime_state, DriverState::INTERNAL_ERROR);
>     }
>     continue;
> }
> ```

I think we can return ok, it won't make a difference. @mofeiatwork what do you think?

Reply:
Maybe we should adopt them both?
1. Cancelled but return OK is little bit strange, in that `OK` is not an error, and would not be thrown to upper caller
2. Cancelled operator should not be processed multiple times, the driver executor should check the `Cancelled` status and handle it as cancelled but not just an error.

Reply:
> Maybe we should adopt them both?
> 
> 1. Cancelled but return OK is little bit strange, in that `OK` is not an error, and would not be thrown to upper caller
> 2. Cancelled operator should not be processed multiple times, the driver executor should check the `Cancelled` status and handle it as cancelled but not just an error.

IMO, cancelled and return OK seems maybe reasonable, because `set_finishing` returning non-OK status means that there is some internal unexpected error, but cancelled is an expected thing.
The upper caller needn't be aware of cancelled status through the returned cancelled status, because it can also be aware of that via `state->is_cancelled()`.

Reply:
I see, your suggestion is return `OK` at `set_finishing`, which is acceptable.

But in general case, there're many other cancelling check points during sort, they need to return a `Cancelled` status to interrupt the sorting procedure, otherwise the state of code is not correct but the procedure continues ,which may introduce some other bugs.

So, as you suggest, we need two approaches to handle the cancel operation, some place return `OK` and let the executor cancel it, and some other places need to return `Cancelled` to interrupt the normal procedure. I'm afraid this could introduce some unnecessary complexity?